File: niu_cloud.py
#!/usr/bin/env python3
"""
niu_cloud.py -- the few NIU cloud calls the scooter CLI needs.

The BLE password that authenticates a phone to a NIU vehicle is not printed
anywhere; the app fetches it from NIU's cloud for a vehicle bound to your
account.  This module logs in the way the Android app does and pulls it:

  POST account-fk.niu.com/v3/api/oauth2/token   account, password=md5(pw), grant_type=password, scope=base, app_id
  GET  app-api-fk.niu.com/v5/scooter/list       header token: <access_token>
  GET  app-api-fk.niu.com/v5/scooter/detail/<sn>
  GET  app-api-fk.niu.com/v5/ble/bleinfo?sn=<sn>   -> bleMac, blePassword, bleAes, bleSign, bleName, bus_protocol_type

Every reply is {"status": 0, "desc": ..., "data": ...}; status != 0 is an error.
"-fk" hosts are the overseas region; the domestic region uses app-api.niu.com.

Credentials live in secrets/ (gitignored), see secrets/README.md.
"""
from __future__ import annotations

import hashlib
import json
import os
import time
import urllib.error
import urllib.parse
import urllib.request

HERE = os.path.dirname(os.path.abspath(__file__))
SECRETS = os.path.join(HERE, "secrets")
SESSION_FILE = os.path.join(SECRETS, "session.json")
SCOOTER_FILE = os.path.join(SECRETS, "scooter.json")

ACCOUNT_HOST = "https://account-fk.niu.com/"
API_HOST = "https://app-api-fk.niu.com/"
APP_IDS = ("niu_8xt1afu6", "niu_ktdrr960")  # overseas release, then domestic release
USER_AGENT = ("manager/5.12.2 (android; Pixel 8 14);lang=en-US;clientIdentifier=Overseas;"
              "timezone=America/Chicago;model=google_Pixel 8;deviceName=Pixel 8;ostype=android")


class CloudError(Exception):
    pass


def _headers(token: str | None = None, json_body: bool = False) -> dict:
    h = {"User-Agent": USER_AGENT, "X-No-Encrypt": "1", "Accept": "application/json"}
    if token:
        h["token"] = token
    if json_body:
        h["Content-Type"] = "application/json"
    return h


def _call(method: str, url: str, token: str | None = None, form: dict | None = None,
          query: dict | None = None, timeout: int = 30) -> dict:
    if query:
        url += ("&" if "?" in url else "?") + urllib.parse.urlencode(query)
    data = urllib.parse.urlencode(form).encode() if form is not None else None
    req = urllib.request.Request(url, data=data, method=method, headers=_headers(token))
    if form is not None:
        req.add_header("Content-Type", "application/x-www-form-urlencoded")
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            body = resp.read().decode("utf-8", errors="replace")
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        try:
            j = json.loads(body)
        except ValueError:
            raise CloudError(f"HTTP {e.code} from {url}: {body[:200]}") from None
        raise CloudError(f"HTTP {e.code}: {j.get('desc') or j.get('message') or body[:200]}") from None
    try:
        j = json.loads(body)
    except ValueError:
        raise CloudError(f"non-JSON reply from {url}: {body[:200]}") from None
    if j.get("status", 0) != 0:
        raise CloudError(f"NIU says: {j.get('desc') or j} (status {j.get('status')})")
    return j.get("data")


# ----------------------------------------------------------------------------- session

def load_json(path: str) -> dict | None:
    try:
        with open(path, encoding="utf-8") as f:
            return json.load(f)
    except FileNotFoundError:
        return None


def save_json(path: str, obj: dict) -> None:
    os.makedirs(os.path.dirname(path), exist_ok=True)
    tmp = path + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(obj, f, indent=1, sort_keys=True)
    os.chmod(tmp, 0o600)
    os.replace(tmp, path)


def login(account: str, password: str, app_id: str | None = None) -> dict:
    """Password login; the app sends md5(password). Tries both app ids unless one is given."""
    last = None
    for aid in ([app_id] if app_id else APP_IDS):
        form = {"account": account, "password": hashlib.md5(password.encode("utf-8")).hexdigest(),
                "grant_type": "password", "scope": "base", "app_id": aid}
        try:
            data = _call("POST", ACCOUNT_HOST + "v3/api/oauth2/token", form=form)
        except CloudError as e:
            last = e
            continue
        tok = (data or {}).get("token") or {}
        if not tok.get("access_token"):
            last = CloudError(f"login reply had no token: {data}")
            continue
        sess = {"account": account, "app_id": aid, "token": tok["access_token"],
                "refresh_token": tok.get("refresh_token", ""),
                "expires_at": int(time.time()) + int(tok.get("token_expires_in") or 0),
                "user": {k: (data.get("user") or {}).get(k) for k in ("id", "name", "email", "mobile", "country_code")}}
        save_json(SESSION_FILE, sess)
        return sess
    raise last or CloudError("login failed")


def refresh(sess: dict) -> dict:
    form = {"refresh_token": sess["refresh_token"], "grant_type": "refresh_token", "scope": "base",
            "app_id": sess.get("app_id", APP_IDS[0])}
    data = _call("POST", ACCOUNT_HOST + "v3/api/oauth2/token", form=form)
    tok = (data or {}).get("token") or {}
    if not tok.get("access_token"):
        raise CloudError(f"refresh reply had no token: {data}")
    sess.update(token=tok["access_token"], refresh_token=tok.get("refresh_token", sess["refresh_token"]),
                expires_at=int(time.time()) + int(tok.get("token_expires_in") or 0))
    save_json(SESSION_FILE, sess)
    return sess


def session() -> dict:
    sess = load_json(SESSION_FILE)
    if not sess or not sess.get("token"):
        raise CloudError("not logged in; run: kqi login you@example.com")
    if sess.get("expires_at", 0) and sess["expires_at"] < time.time() + 3600 and sess.get("refresh_token"):
        try:
            sess = refresh(sess)
        except CloudError:
            pass
    return sess


# ----------------------------------------------------------------------------- vehicle data

def scooters(token: str) -> list[dict]:
    data = _call("GET", API_HOST + "v5/scooter/list", token=token)
    if isinstance(data, dict):
        data = data.get("items") or data.get("list") or []
    return data or []


def detail(token: str, sn: str) -> dict:
    return _call("GET", API_HOST + "v5/scooter/detail/" + urllib.parse.quote(sn), token=token) or {}


def bleinfo(token: str, sn: str) -> dict:
    return _call("GET", API_HOST + "v5/ble/bleinfo", token=token, query={"sn": sn}) or {}


def load_scooter() -> dict:
    sc = load_json(SCOOTER_FILE)
    if not sc or not sc.get("ble", {}).get("mac"):
        raise CloudError("no scooter credentials; run: kqi setup")
    return sc


def save_scooter(sc: dict) -> None:
    save_json(SCOOTER_FILE, sc)

File: ios/KQiRides/Sources/NIUCloud.swift
import Foundation
import CryptoKit

/// NIU cloud client: the few calls this app needs, ported from `niu_cloud.py`.
///
/// The overseas region is the "-fk" host pair. Every reply is
/// `{status, desc, data}` and a non-zero status is an error, including on HTTP 200.
actor NIUCloud {
    static let shared = NIUCloud()

    private let accountHost = "https://account-fk.niu.com/"
    private let apiHost = "https://app-api-fk.niu.com/"
    private let appIDs = ["niu_8xt1afu6", "niu_ktdrr960"]

    /// The app identifies as the Android client; the API is picky about the shape
    /// of this string. The timezone here does NOT drive the ride-date bug: the
    /// backend buckets days in its own zone regardless of what we claim.
    private var userAgent: String {
        "manager/5.12.2 (android; Pixel 8 14);lang=en-US;clientIdentifier=Overseas;"
        + "timezone=\(TimeZone.current.identifier);model=google_Pixel 8;deviceName=Pixel 8;ostype=android"
    }

    struct CloudError: LocalizedError {
        let msg: String
        var errorDescription: String? { msg }
    }

    private func request(_ method: String, _ url: URL, token: String?,
                         form: [String: String]? = nil, json: [String: Any]? = nil) async throws -> Any {
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        req.setValue("1", forHTTPHeaderField: "X-No-Encrypt")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token { req.setValue(token, forHTTPHeaderField: "token") }
        if let form {
            req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
            var c = URLComponents()
            c.queryItems = form.map { URLQueryItem(name: $0.key, value: $0.value) }
            req.httpBody = c.percentEncodedQuery?.data(using: .utf8)
        } else if let json {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONSerialization.data(withJSONObject: json)
        }
        let (data, _) = try await URLSession.shared.data(for: req)
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw CloudError(msg: "non-JSON reply from \(url.lastPathComponent)")
        }
        let status = (obj["status"] as? Int) ?? 0
        guard status == 0 else {
            throw CloudError(msg: (obj["desc"] as? String) ?? "cloud error \(status)")
        }
        return obj["data"] ?? [:]
    }

    // MARK: - auth

    struct Session: Codable {
        var account: String
        var appID: String
        var token: String
        var refreshToken: String
        var expiresAt: Date
    }

    func login(account: String, password: String) async throws -> Session {
        let md5 = Insecure.MD5.hash(data: Data(password.utf8)).map { String(format: "%02x", $0) }.joined()
        var last: Error = CloudError(msg: "login failed")
        for aid in appIDs {
            do {
                let data = try await request("POST", URL(string: accountHost + "v3/api/oauth2/token")!,
                    token: nil,
                    form: ["account": account, "password": md5, "grant_type": "password",
                           "scope": "base", "app_id": aid])
                guard let d = data as? [String: Any], let tok = d["token"] as? [String: Any],
                      let access = tok["access_token"] as? String, !access.isEmpty else {
                    last = CloudError(msg: "login reply had no token"); continue
                }
                let ttl = (tok["token_expires_in"] as? Double) ?? 0
                return Session(account: account, appID: aid, token: access,
                               refreshToken: (tok["refresh_token"] as? String) ?? "",
                               expiresAt: Date().addingTimeInterval(ttl))
            } catch { last = error }
        }
        throw last
    }

    func refresh(_ s: Session) async throws -> Session {
        let data = try await request("POST", URL(string: accountHost + "v3/api/oauth2/token")!,
            token: nil,
            form: ["refresh_token": s.refreshToken, "grant_type": "refresh_token",
                   "scope": "base", "app_id": s.appID])
        guard let d = data as? [String: Any], let tok = d["token"] as? [String: Any],
              let access = tok["access_token"] as? String else {
            throw CloudError(msg: "refresh reply had no token")
        }
        var out = s
        out.token = access
        out.refreshToken = (tok["refresh_token"] as? String) ?? s.refreshToken
        out.expiresAt = Date().addingTimeInterval((tok["token_expires_in"] as? Double) ?? 0)
        return out
    }

    // MARK: - vehicle

    func scooters(token: String) async throws -> [Scooter] {
        let data = try await request("GET", URL(string: apiHost + "v5/scooter/list")!, token: token)
        let items: [Any]
        if let arr = data as? [Any] { items = arr }
        else if let d = data as? [String: Any] { items = (d["items"] as? [Any]) ?? (d["list"] as? [Any]) ?? [] }
        else { items = [] }
        let raw = try JSONSerialization.data(withJSONObject: items)
        return (try? JSONDecoder().decode([Scooter].self, from: raw)) ?? []
    }

    struct BLEInfo: Codable {
        let bleMac: String?
        let bleName: String?
        let blePassword: String?
        let bleAes: String?
        let bleSign: String?
    }

    /// The BLE credentials for a scooter, by serial. This is what lets an iPhone
    /// talk to the scooter at all: CoreBluetooth hides MAC addresses, so the
    /// macOS-only MAC discovery the CLI uses is not available here.
    func bleInfo(token: String, sn: String) async throws -> BLEInfo {
        var c = URLComponents(string: apiHost + "v5/ble/bleinfo")!
        c.queryItems = [URLQueryItem(name: "sn", value: sn)]
        let data = try await request("GET", c.url!, token: token)
        let raw = try JSONSerialization.data(withJSONObject: data)
        return try JSONDecoder().decode(BLEInfo.self, from: raw)
    }

    func detail(token: String, sn: String) async throws -> [String: Any] {
        let data = try await request("GET", URL(string: apiHost + "v5/scooter/detail/" + sn)!, token: token)
        return (data as? [String: Any]) ?? [:]
    }

    // MARK: - rides

    /// One page of rides. `index` is a 1-based page number the server insists on
    /// receiving as a *string* while `pagesize` must be a number; sending both as
    /// the same type fails parameter validation. v3 is the endpoint that answers
    /// for kick scooters (v2 accepts the call but always returns an empty list).
    func rides(token: String, sn: String, page: Int, pageSize: Int = 20) async throws -> [Ride] {
        let data = try await request("POST", URL(string: apiHost + "v5/track/list/v3")!, token: token,
                                     json: ["sn": sn, "pagesize": pageSize, "index": String(page)])
        guard let d = data as? [String: Any], let items = d["items"] as? [Any] else { return [] }
        let raw = try JSONSerialization.data(withJSONObject: items)
        return (try? JSONDecoder().decode([Ride].self, from: raw)) ?? []
    }

    /// Every ride, walking pages until one comes back empty.
    func allRides(token: String, sn: String, maxPages: Int = 50) async throws -> [Ride] {
        var out: [Ride] = []
        var seen = Set<String>()
        for p in 1...maxPages {
            let batch = try await rides(token: token, sn: sn, page: p)
            let fresh = batch.filter { !seen.contains($0.trackId) }
            if fresh.isEmpty { break }
            fresh.forEach { seen.insert($0.trackId) }
            out += fresh
        }
        return out.sorted { $0.startTime > $1.startTime }
    }

    // MARK: - firmware

    static let otaDeviceTypes = ["FOC", "DB", "BMS", "LCU", "ECU_BT"]

    func firmware(token: String, sn: String) async throws -> [[String: Any]] {
        let devices = Self.otaDeviceTypes.map {
            ["devicetype": $0, "soft_version": "0.0.0", "hard_version": "0.0.0"]
        }
        let data = try await request("POST", URL(string: apiHost + "v5/ota/checkupdate")!, token: token,
                                     json: ["sn": sn, "devices": devices])
        return ((data as? [String: Any])?["items"] as? [[String: Any]]) ?? []
    }
}
